src/jit/codegenxarch.cpp

```diff
@@ -7454,13 +7454,16 @@ CodeGen::genIntToFloatCast(GenTreePtr treeNode)
 
     // Since xarch emitter doesn't handle reporting gc-info correctly while casting away gc-ness we
     // ensure srcType of a cast is non gc-type.  Codegen should never see BYREF as source type except
-    // for GT_LCL_VAR_ADDR and GT_LCL_FLD_ADDR that represent stack addresses and can be considered
-    // as TYP_I_IMPL. In all other cases where src operand is a gc-type and not known to be on stack,
+    // for GT_LCL_VAR_ADDR, GT_LCL_FLD_ADDR and GT_LCL_VAR with lvIsVarAddrOnStack
+    // that represent stack addresses and can be considered as TYP_I_IMPL.
+    // In all other cases where src operand is a gc-type and not known to be on stack,
     // Front-end (see fgMorphCast()) ensures this by assigning gc-type local to a non gc-type
     // temp and using temp as operand of cast operation. 
     if (srcType == TYP_BYREF)
     {
-        noway_assert(op1->OperGet() == GT_LCL_VAR_ADDR || op1->OperGet() == GT_LCL_FLD_ADDR);
+        noway_assert(op1->OperGet() == GT_LCL_VAR_ADDR || op1->OperGet() == GT_LCL_FLD_ADDR
+            || (op1->OperGet() == GT_LCL_VAR &&
+                !!compiler->lvaTable[op1->gtLclVarCommon.GetLclNum()].lvIsVarAddrOnStack));
```

Comment:
Please not the above comment, if it is anything other than lclvar_addr or lcl_fld_addr, then fgMorphCast() should transform IR as follows whenever casting away gc-ness

TYP_I_IMPL  tmp = lcl var of type byref
cast (double <-byref, tmp)

We need to see why fgMorphCast() is failing to make this transformation and if that logic is missing we need to add it.